x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizer.java

```diff
@@ -156,19 +111,22 @@ public PhysicalPlan apply(PhysicalPlan plan) {
                                 output = Expressions.asAttributes(fields);
                             }
                             // add a logical projection (let the local replanning remove it if needed)
-                            p = exec.replaceChild(
-                                new FragmentExec(
-                                    Source.EMPTY,
-                                    new Project(logicalFragment.source(), logicalFragment, output),
-                                    fragmentExec.esFilter(),
-                                    fragmentExec.estimatedRowSize(),
-                                    fragmentExec.reducer()
-                                )
+                            FragmentExec newChild = new FragmentExec(
+                                Source.EMPTY,
+                                new Project(logicalFragment.source(), logicalFragment, output),
+                                fragmentExec.esFilter(),
+                                fragmentExec.estimatedRowSize(),
+                                fragmentExec.reducer()
                             );
+                            return new ExchangeExec(exec.source(), output, exec.isInBetweenAggs(), newChild);
                         }
                     }
+                } else {
+                    AttributeSet childOutput = currentPlanNode.inputSet();
+                    AttributeSet addedAttributes = currentPlanNode.outputSet().subtract(childOutput);
+                    requiredAttributes.set(requiredAttributes.get().subtract(addedAttributes).combine(currentPlanNode.requiredInputSet()));
```

Comment:
The main change; the way we collected all attributes that occurred and then removed generated attributes is fully abstracted away here.